Merge @zostay mutator work

jonathanstowe opened the pull request:
This is to cover off #27 fully rebased and all tests working.

I wouldn't recommend merging this unless it's tested with some modules that use it (i.e. things that can't be done in the unit tests.)

zostay committed:
WIP adding mutators to URI

zostay committed:
Move away from accessor/mutator interface

Also add path modification

zostay committed:
Adding basic query and query-form manipulation

Also making query-form immutable

zostay committed:
Testing that %!query-form is NOT mutable

zostay committed:
Make fragment rw

zostay committed:
$!is_absolute is set but never used, deleting

zostay committed:
Making authority a full blown class

zostay committed:
Upgrade URI::Path from subset to class

zostay committed:
Making URI::Query a full-blown object

This implementation seems like a good compromise to me between simplicity and
flexibility.

zostay committed:
Return Nil from _port when no authority

zostay committed:
Tightening up the API

zostay committed:
Further tightening up of the API

Deprecating query-form, it is no longer necessary

zostay committed:
Expanding the POD

zostay committed:
Making authority components return '' when not set

Also more docs

zostay committed:
More docs, outlined what remains

zostay committed:
More docs

Also I deleted Capture here because I believe there's no unambiguous coercion to
Capture for the URI::Query object. At the very least, we would want to define it
so that it takes hash-format into consideration and trying to think what that
would look like gave me a headache so I just dropped it instead.

zostay committed:
authors, copyright, and license

zostay committed:
bump version, set auth

zostay committed:
Fix provides

zostay committed:
Adding Test::META test

zostay committed:
Adding a test for URI::Query

zostay committed:
A few more URI::Query tests

zostay committed:
Add test for URI::Authority

zostay committed:
Found a flaw in my understanding of authority

must fix

zostay committed:
Starting to fix my flawed understanding of authority

zostay committed:
Fix the final test

zostay committed:
More precise path and segment handling

zostay committed:
Tightening up the docs

Also allow :hash-format and :!hash-format boolean flags to split-query().

zostay committed:
Fix travis-ci failing because no deps install

zostay committed:
update README.md

jonathanstowe committed:
Naive re-base

All tests passing

jonathanstowe committed:
Standardize the return declarations

jonathanstowe committed:
Remove Test::META as a test dependency

The meta test is optional on the presence of the module anyway.

jonathanstowe committed:
Oh two META tests

One which skips if the module isn't there and one which doesn't

jonathanstowe commented:
It's still not passing on Appveyor but I'm not sure what that's about.

jonathanstowe committed:
Remove ambiguous query-form candidate

Upgraded deprecation warnings to hard deprecations

jonathanstowe committed:
Fix a warning with no path segments

jonathanstowe committed:
Possibly vain attempt to fix appveyor

Scimon reviewed (approved):
So there's a *lot* here but it all looks sane (there were a couple of methods it took me a moment to twig (positional vs named basically)). 
The test suite looks complete so I would err on the side of merging it. If there are bugs they will be edge cases we can handle in the future.

Scimon commented on appveyor.yml:
```diff
@@ -17,6 +17,7 @@ install:
   - git clone https://github.com/tadzik/rakudobrew %USERPROFILE%\rakudobrew
   - SET PATH=%USERPROFILE%\rakudobrew\bin;%PATH%
   - rakudobrew build moar %TEST_MOAR%
+  - SET PATH=%USERPROFILE%\rakudobrew\versions\moar-%TEST_MOAR%\install\bin:%PATH%
```
This is the issues with appveyor failing I think should be 

SET PATH=%USERPROFILE%\rakudobrew\versions\moar-%TEST_MOAR%\install\bin;%PATH%

jonathanstowe replied:
Hahah yeah cheers.

jonathanstowe replied:
Well that definitely moved on the appveyor to a different fail in zef.

Scimon replied:
Yeah. Might it be the Moar version?

Scimon reviewed (changes requested):
The rest of my review stands but I think this is why Appveyor is faling. Windows path seperator.

jonathanstowe committed:
Fix oopsie with windows path separator

From @scimon

Scimon commented:
I think the test-moar: '' doesn't work because it tries to set the path to moar-/bin to get perl6.
I think you need to set the version. (Also the other is runing on 2017-05. I'm not suprised if zef won't install.

jonathanstowe commented:
Eek is there any way to set the version to that which rakudobrew  just built?

jonathanstowe commented:
Okay on the grounds that no-one has touched the appveyor config since it was added and it has been failing consistently I'm going to remove it.  If anyone feels strongly enough they can make it work again.

jonathanstowe committed:
Remove appveyor config

It doesn't work right now and hasn't for a while.
Feel free to put back a working one.

jonathanstowe commented:
I'm going to merge this manually to up version, Changes, etc easier.

jonathanstowe commented:
Merged with abe8c9bb65947760cb656c6c154f466cd87f6e57  as v0.3.0
